pkg/apis/core/v1beta1/types_shoot.go

```diff
@@ -741,6 +741,8 @@ type Alerting struct {
 // Provider contains provider-specific information that are handed-over to the provider-specific
 // extension controller.
 type Provider struct {
+	// EnableContainerD indicate if kubelet should be started with ContainerD and not docker
+	EnableContainerD bool  `json:"enableContainerD,omitempty"`
```

Comment:
Shouldn't this rather be part of the `Worker` struct so that one can enable/disable it per worker pool?
Also, WDYT about using a struct instead of just a bool? This would help in the future in case we decide to allow specifying more configuration options:

```golang
type ContainerD struct {
  // Enabled specifies whether containerd should be enabled or not.
  Enabled bool `json:"enabled"`
}
```

...

```golang
type Worker struct {
  ...
  // ContainerD contains configuration settings for containerd.
  // +optional
  ContainerD *ContainerD `json:"containerd,omitempty"`
}
```

Reply:
Ok, sure. I first put it under worker but noticed i'm changing it only by Shoot logic (Hibernation turns on, Kuberenetes version update etc) I didn't find a scenario to change it per Worker.
If will be possible for users to change it per worker but i'm not sure this is relevant because this change should be "transparent" to their cluster.
If you think it is does make sense to set it per Worker, maybe we can create a new struct which will already be used for the container runtime as well.
`
type ContainerRuntime struct {
  // Enabled specifies whether containerd should be enabled or not.
  ContainerDEnabled bool `json:"enabled"`
  GVisorEnabled.        bool `json:"enabled"`
}
`
WDYT?

Reply:
I don't have a scenario either, but that doesn't mean that no one exists. If there is no reason to not make it controllable per worker pool then we usually always prefer to give the end-users the maximum possible flexibility. Hence, I'd vote for moving it there.

> If you think it is does make sense to set it per Worker, maybe we can create a new struct which will already be used for the container runtime as well.

Hm, we should not "hard-code" the possible container runtimes in the API. It should be like you proposed in the GEP10, so a list of container runtimes plus proper provider configs that are not understandable by Gardener.

Reply:
True. it was here by mistake, this should be:

type Worker struct {
  ...
  // ContainerD contains configuration settings for containerd.
  // +optional
  Cri *Cri `json:"cri,omitempty"`
}

type Cri struct {
  // Enabled specifies whether containerd should be enabled or not.
  ContainerDEnabled bool `json:"enabled"`
 ContainerRuntime   []containerRuntime `json:"containerRuntime"`
}

type ContainerRuntime struct {
 type string  `json:"type"`
}

Reply:
Ah, ok, understood. Yes, this looks great, it makes sense 👍 thank you

Reply:
BTW, I've heard about users interacting with the docker socket on the nodes.
So the fact that we are not aware about some use-case does not mean that it does not exist.

And I am against changing the behavior in the current version of the API or forcefully migrating existing shoot clusters.